### PR TITLE
move to documentation website

### PR DESCRIPTION
**Pull request by jaxxzer**
Note actual commit topology in this pr is very different from github display order, it might help to do the review in `tig`.

### CODE REVIEW

**Comment by jaxxzer** [comment]
Documentation website preview: http://docs.bluerobotics.com/ping-protocol/

**Comment by jaxxzer** [comment]
@patrickelectric @Williangalvani I can't figure out why travis stopped running, I need help. Feel free to force push.

**Comment by jaxxzer** [comment]
resolved, thanks.

**Comment by jaxxzer** [comment]
Next steps:
- Check out the new [ping-cpp](https://github.com/bluerobotics/ping-cpp) repository (including the [deployment](https://github.com/bluerobotics/ping-cpp/tree/deployment) branch)
- review the ping-viewer pr to move to ping-cpp submodule:
- I will soon submit a ping-arduino pr to use ping-cpp submodule in the same manner as ping-viewer ( I will make arduino work with the submodule, don't worry about that)
- I will soon submit a ping-python pr to migrate the test/ci/deployment that was inside of the ping-protocol repository into the ping-python repository (make ping-python have ci and structure that is similar to ping-cpp repository)

**Comment by jaxxzer** [comment]
Here is an image that describes the repository layout I envision at the end of this work: 
ping-protocol-planning - New frame (2)

**Comment by jaxxzer** [comment]
I've recorded a few advantages we will see with this work:
- ready-to-checkout source code is available in implementation repositories after checkout of deployment branch (no need to run python locally) (ping viewer no longer depends on python)
- we now have an independent minimal cpp library, that has no dependencies
- other dependency reductions, simplified ci scripts, and faster ci times
- NO AUTOGENERATED content is tracked in development branch (no more forgetting to update protocol readme and failing ci)
- arduino and ping viewer share parser code (both previous parsers had bugs!)
- generator class is improved with generate method, and generation scripts and templates are reduced and simplified
- development and deployment is appropriately compartmentalized:
  - deployment of implementation repositories is independent of commits in ping-protocol (no more broken implementations)
  - python files moved to python repo, cpp files moved to cpp repo, development of implementation apis is reflected in implementation repository history (instead of remote ping-protocol repository)
- test cases are made by hand instead of generated (more robust):
  - static test cases (not autogenerated) mean that if we break the api, we will know (ex. when refactoring class name PingMessage -> ping_message)
  - automatically generated test cases risk bugs in the implementation (bug in the code generating the test case may cause actual problems to go undetected)

File: .travis.yml [hunk 1]
```diff
@@ -1,73 +1,26 @@
...
 script:
-- if [ "${TRAVIS_OS_NAME}" = "linux" ]; then
-    sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-7 60;
-    sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-7 60;
-    source /opt/qt*/bin/qt*-env.sh;
-  fi
-- if [ "${PINGTEST}" = "true" ]; then
-    tools/run-tests.sh || travis_terminate 1;
-    tools/update-generated-repositories.sh;
+- |
```
Comment (patrickelectric) [inline .travis.yml]: I think this should be in a script inside like the old version, there is no reason to be here.
Add this kind of stuff inside yaml makes it harder to maintain and test.

File: src/protocol/generate-markdown.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    os.makedirs(args.output_directory)
+
+scriptPath = Path(__file__).parent.absolute()
+
+definitionPath = "%s/definitions" % scriptPath
+templatePath = "%s/templates" % scriptPath
+
+templateFile = "%s/pingmessage-.md.in" % templatePath
+
+g = Generator()
```
Comment (patrickelectric) [inline src/protocol/generate-markdown.py]: Do not use single letter variables

File: src/protocol/generator.py [hunk 3]
```diff
@@ -1,17 +1,14 @@
...
-import os
 import json
 import re
+import collections
+from jinja2 import Template
+from pathlib import Path
 
-from jinja2 import Environment, FileSystemLoader
-
+# TODO the generator api should help with determining things about messages (like a total payload length, or if a field is a vector type)
```
Comment (patrickelectric) [inline src/protocol/generator.py]: Avoid long lines with comments, break it in more lines.

File: src/protocol/generator.py [hunk 4]
```diff
@@ -52,7 +49,7 @@ def get_c_size(self, t):
         # this regex will get the X in u/intX_t (uint8_t, int16_t)
         match = re.search('[0-9]{1,2}', t)
         if match:
-            return int(match.group(0)) / 8
+            return int(int(match.group(0)) / 8)
```
Comment (patrickelectric) [inline src/protocol/generator.py]: why is this necessary ?

File: src/protocol/generator.py [hunk 5]
```diff
@@ -101,44 +98,25 @@ def get_type_string(self, t, pointer=False, name=''):
     def is_vector(self, t):
         return ('vector' in t)
 
-    def capitalize(self, input_string):
-        capitalized_words = [ x[0].capitalize() + x[1:] for x in input_string.split('_')]
-        return ''.join(capitalized_words)
+    def generate(self, definition, template, globies=None):
+        d = json.load(open(definition), object_pairs_hook=collections.OrderedDict)
```
Comment (patrickelectric) [inline src/protocol/generator.py]: Avoid single letter variables.

File: src/protocol/generator.py (hunk 3 above)
Comment (patrickelectric) [inline src/protocol/generator.py]: Use docstrings: https://pythonhelp.wordpress.com/2011/02/14/docstrings/

File: src/protocol/templates/pingmessage_.h.in [hunk 6]
```diff
@@ -9,47 +9,48 @@
 
 #include "pingmessage.h"
 
-{% set message_type = _actual_message_type %}
-{% for message in messages[message_type]['all_msgs']|sort %}
-{% set total_payload = generator.calc_payload(messages[message_type]['all_msgs'][message].payload) %}
+{% for message_type in messages %}
+{% for message in messages[message_type] %}
+{% set m = messages[message_type][message] %}
```
Comment (patrickelectric) [inline src/protocol/templates/pingmessage_.h.in]: Avoid single letter variables

File: .travis.yml (hunk 1 above)
Comment (jaxxzer) [reply .travis.yml]: Agreed.

File: src/protocol/generator.py (hunk 4 above)
Comment (jaxxzer) [reply src/protocol/generator.py]: Previously, only the dividend was converted, and the return was still float. My fix here was not the best (just needed to move a parenthesis). I'll take care of it.

Comment (jaxxzer) [reply src/protocol/generator.py]: (We were getting things like `msgData[2.0]`, this patch fixes that)

File: src/protocol/templates/pingmessage_.h.in (hunk 6 above)
Comment (jaxxzer) [reply src/protocol/templates/pingmessage_.h.in]: (this file has been removed, I checked the others)

**Comment by jaxxzer** [comment]
Did a few more things this morning:
- removed the stale templates
- moved `src/protocol/*` to `just src/*`
- updated gitignore
now the repository is totally clean: https://github.com/jaxxzer/ping-protocol/tree/doc

**Comment by jaxxzer** [comment]
@patrickelectric done

File: tools/travis-ci-script.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+
+for filename in src/definitions/*.json; do
+echo "Checking file: ${filename}"
+python3 -m json.tool ${filename} > /tmp/temporary_test_file.json;
+if ! comm -2 -3 ${filename} /tmp/temporary_test_file.json; then
+    echo "Json file does not follow style. Try python3 -m json.tool."
+    exit 1
+fi
+done
```
Comment (patrickelectric) [inline tools/travis-ci-script.sh]: It's missing the indentation here.

File: src/generator.py [hunk 8]
```diff
@@ -101,44 +102,26 @@ def get_type_string(self, t, pointer=False, name=''):
     def is_vector(self, t):
         return ('vector' in t)
 
-    def capitalize(self, input_string):
-        capitalized_words = [ x[0].capitalize() + x[1:] for x in input_string.split('_')]
-        return ''.join(capitalized_words)
+    def generate(self, definition, template, globuls=None):
+        definitionDict = json.load(open(definition), object_pairs_hook=collections.OrderedDict)
```
Comment (Williangalvani) [inline src/generator.py]: while it will probably work fine, using `open()` like this leaves the file opened (ppparently until is is garbage-collected).

I'd rather use the [`with open(..)`](https://docs.python-guide.org/writing/style/#read-from-a-file) syntax or at least explicitly close the file.

File: src/generator.py [hunk 9]
```diff
@@ -1,17 +1,17 @@
 #!/usr/bin/env python3
 
-import os
 import json
 import re
-
-from jinja2 import Environment, FileSystemLoader
+import collections
+from jinja2 import Template
+from pathlib import Path
```
Comment (Williangalvani) [inline src/generator.py]: Is this used?

File: src/generator.py [hunk 10]
```diff
@@ -101,44 +102,26 @@ def get_type_string(self, t, pointer=False, name=''):
     def is_vector(self, t):
         return ('vector' in t)
 
-    def capitalize(self, input_string):
-        capitalized_words = [ x[0].capitalize() + x[1:] for x in input_string.split('_')]
-        return ''.join(capitalized_words)
+    def generate(self, definition, template, globuls=None):
```
Comment (Williangalvani) [inline src/generator.py]: What is a "globul"? should it be "glob"?

File: src/generator.py [hunk 11]
```diff
@@ -52,7 +53,7 @@ def get_c_size(self, t):
         # this regex will get the X in u/intX_t (uint8_t, int16_t)
         match = re.search('[0-9]{1,2}', t)
         if match:
-            return int(match.group(0)) / 8
+            return int(match.group(0) / 8)
```
Comment (Williangalvani) [inline src/generator.py]: Does it work?
I expected the result of  `match.group()` to be a string, which should not be divisible by 8.

File: src/generator.py (hunk 10 above)
Comment (jaxxzer) [reply src/generator.py]: `global` is a python keyword

File: src/generator.py (hunk 11 above)
Comment (jaxxzer) [reply src/generator.py]: it does with python3 :) and travis ci

Comment (jaxxzer) [reply src/generator.py]: > I expected the result of match.group() to be a string, which should not be divisible by 8.

though, that is probably why I found it the way it was. We (I) started development of this repo with python 2.

File: tools/travis-ci-script.sh [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+
+for filename in src/definitions/*.json; do
+echo "Checking file: ${filename}"
+python3 -m json.tool ${filename} > /tmp/temporary_test_file.json;
```
Comment (patrickelectric) [inline tools/travis-ci-script.sh]: still missing indentation.

File: src/generator.py (hunk 9 above)
Comment (jaxxzer) [reply src/generator.py]: nope!

File: src/generator.py (hunk 8 above)
Comment (jaxxzer) [reply src/generator.py]: done

File: tools/travis-ci-script.sh (hunk 12 above)
Comment (jaxxzer) [reply tools/travis-ci-script.sh]: XXXXXXXX
